src/modes/run.ts

```diff
@@ -1,5 +1,340 @@
-import { RunOptions } from "../types";
+import { RunOptions, SingularRunOptions } from "../types";
+import { ChainContext, Registry, ReplayPlan } from "../types/model";
+import DBService from "../utils/db";
+import { ComposableCoW, ComposableCoW__factory } from "../types/generated";
+import { COMPOSABLE_COW_CONTRACT_ADDRESS } from "@cowprotocol/cow-sdk";
+import { ConditionalOrderCreatedEvent } from "../types/generated/ComposableCoW";
+import { addContract } from "../addContract";
+import { checkForAndPlaceOrder } from "../checkForAndPlaceOrder";
 
-export async function run(_options: RunOptions) {
-  throw new Error("Not implemented");
+/**
+ * Run the watch-tower 👀🐮
+ * @param options Specified by the CLI / environment for running the watch-tower
+ */
+export async function run(options: RunOptions) {
+  const { rpc, deploymentBlock, oneShot } = options;
+
+  process.on("unhandledRejection", async (error) => {
+    console.log(error);
+    await DBService.getInstance().close();
+    process.exit(1);
+  });
+
+  process.on("SIGINT", async function () {
+    console.log("️⚠️ Caught interrupt signal ⚠️");
+    await DBService.getInstance().close();
+    process.exit(0);
+  });
+
+  try {
+    const chainWatchers = await Promise.all(
+      rpc.map((rpc, index) => {
+        return ChainWatcher.init(
+          {
+            ...options,
+            rpc,
+            deploymentBlock: deploymentBlock[index],
+          },
+          DBService.getInstance()
+        );
+      })
+    );
+
+    // Run the block watcher for each chain
+    const runPromises = chainWatchers.map(async (chainWatcher) => {
+      return chainWatcher.warmUp(oneShot);
+    });
+
+    // Run all the chain watchers
+    await Promise.all(runPromises);
+  } catch (error) {
+    console.error(error);
+    await DBService.getInstance().close();
+    process.exit(1);
+  } finally {
+    await DBService.getInstance().close();
+  }
```

Comment:
If I'm not mistaken, finally will only run in a regular execution, since catch exits before it, no?
Either way, the DB is already closed in that case.
Might just move line `55` after `49` and remove the finally

Reply:
Fixed.